feat(profiling): add support to ingest non-sampled profiles

viglia opened the pull request:
This will let us use data of non sampled profiles to enhance current aggregation (slowest function) with more data.

Currently under feature flag for experimentation with the `Sentry` org.


Feature flag definition `getsentry` [here](https://github.com/getsentry/getsentry/pull/11335) and `sentry` [here](https://github.com/getsentry/sentry/pull/54023).

⚠  Please, do not merge this before the appropriate `vroom` [PR](https://github.com/getsentry/vroom/pull/300) is merged.

Commit:
Add support (feature flagged) to ingest non-sampled profiles

This will let us use data of non sampled profiles to enhance current
aggregation (slowest function) with more data.

Commit:
Remove leftover comment

Commit:
Fix errors

Zylphrex commented on relay-server/src/actors/processor.rs:
```diff
@@ -1066,8 +1066,18 @@ impl EnvelopeProcessorService {
+                        .unwrap_or(true) =>
+            {
+                ItemAction::DropSilently
+            }
```
We **never** want to keep any profiles that came in their own envelope. This includes the old cocoa SDK which we do not want to support. So let's make sure to scope this to just profiles that came in the same envelope as the corresponding transaction.

viglia replied:
@Zylphrex for profiles that get sampled we still do this check.

We only do this for profiles that were non sampled due to `DS`.

@jjbayer was explaining to me (IIUC) that there is a chain of operations on the envelope as it gets processed, so if a envelope comes with transaction + profile, and then the transaction is not sampled due to DS, then the output would be the envelope without the transaction.

Gotta check with @jjbayer if there is another way to check wether the `transaction_count==0` because of DS or because it was not sent over to begin with.

phacops replied:
This `transaction_count` is counting the number of transaction in the envelope as it was received. So, if it's 0, it's not because of DS and we want to discard the profile no matter what.

Commit:
Undo transaction_count changes

jjbayer commented on relay-server/src/actors/processor.rs:
```diff
@@ -2346,14 +2346,28 @@ impl EnvelopeProcessorService {
+                let outcome = Outcome::FilteredSampling(rule_ids.clone());
+                state.managed_envelope.retain_items(|item| {
+                    if unsampled_profiles_enabled && item.ty() == &ItemType::Profile {
+                        item.set_header("sampled", false);
```
To reduce processing cost, could we skip [`process_profiles`](https://github.com/getsentry/relay/blob/13d9953162673ab546f10108d3bd87ecba718602/relay-server/src/actors/processor.rs#L1121) for profiles with `sampled: false`?

viglia replied:
@jjbayer I don't think so. 

If the `feature-flag` is enabled for a given org, and we want to send non-sampled profiles over, we still need to check that those are proper profiles (respect max size, profile payload properly formed, etc.)

jjbayer reviewed (commented):
We definitely need some tests for this, ideally with a chain of relays to see if the item header is propagated correctly. See for example

https://github.com/getsentry/relay/blob/13d9953162673ab546f10108d3bd87ecba718602/tests/integration/test_dynamic_sampling.py#L560-L590

jjbayer committed:
fix(server): Allow partial update of envelope summary during processing (#2383)

Updating the envelope summary during envelope processing is currently a
no-go, because recomputing the summary will miss the event item, which
is taken from the envelope before processing.

https://github.com/getsentry/relay/pull/2380 contains an attempt to
remove caching of the envelope summary entirely. As a quick fix, this PR
provides a partial update function.

#skip-changelog

Commit:
Merge branch 'master' into viglia/feat/ingest-non-sampled-profiles

Commit:
Add debug test_relay_chain_keep_unsampled_profile

Commit:
Debug dropping event

Commit:
Add transactionMetrics config

Commit:
Add options that we pass down from Sentry
